lib/streamlit/runtime/caching/cached_message_replay.py

```diff
@@ -260,19 +254,36 @@ def calling_cached_function(
         It allows us to track any `st.foo` messages that are generated from inside the function
         for playback during cache retrieval.
         """
-        self._cached_func_stack.append(func)
         self._cached_message_stack.append([])
         self._seen_dg_stack.append(set())
-        if allow_widgets:
-            self._allow_widgets += 1
+        self._allow_widgets = allow_widgets
+
+        nested_call = False
+        ctx = get_script_run_ctx()
+        if ctx:
+            if ctx.disallow_cached_widget_usage:
+                # The disallow_cached_widget_usage is already set to true.
+                # This indicates that this cached function run is called from another
+                # cached function that disallows widget usage.
+                # We need to deactivate the widget usage for this cached function run
+                # even if it was allowed.
+                self._allow_widgets = False
+                nested_call = True
+
+            if not self._allow_widgets:
+                # If we're in a cached function that disallows widget usage, we need to set
+                # the disallow_cached_widget_usage to true for this cached function run
+                # to prevent widget usage (triggers a warning).
+                ctx.disallow_cached_widget_usage = True
```

Comment:
Could you move this into a convenient function `_check_allow_widgets` or similar and leave a docstring comment explaining the behavior together with a small example? I think that might make it cleaner and easier to understand what's going on here

Reply:
Hmm, the reason why it looks long is that it contains a lot of commentary. But the actual code logic that I could put into another function is only a couple of lines. E.g. this is the full code logic of the function:

<img width="706" alt="image" src="https://github.com/streamlit/streamlit/assets/2852129/8414909e-e0e8-4c37-aaed-40cd6e53cdd4">

I don't think its worth doing that here since I need to modify class variables + share variables with other parts of the code lie `ctx` and `self._allow_widgets`